pkg/acpi/coreboot/madt.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package coreboot
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/u-root/u-root/pkg/acpi"
+)
+
+const (
+	minMADT = 8 // 32-bit address and flags
+)
+const (
+	entryLAPIC         = 0
+	entryIOAPIC        = 1
+	entryISOR          = 2 // pronounced: eyesore, for reasons.
+	entryBOGUS         = 3
+	entryNMI           = 4
+	entryLAPICOverRide = 5
+	entryLX2APIC       = 9
+	entryLX2NMI        = 10
+
+	sizeEntryLAPIC         = 8
+	sizeEntryIOAPIC        = 12
+	sizeEntryISOR          = 10 // pronounced: eyesore, for reasons.
+	sizeEntryNMI           = 6
+	sizeEntryLAPICOverRide = 12
+	sizeEntryLX2APIC       = 16
+	sizeEntryLX2NMI        = 12
+)
+
+var (
+	// The MADT signature is APIC.
+	// Another failure of vision; they never thought there'd
+	// be more than one APIC I guess.
+	sigMADT = "APIC"
+	info    = map[int]struct {
+		name string
+		size int
+	}{
+		entryLAPIC:         {"LAPIC", sizeEntryLAPIC},
+		entryIOAPIC:        {"IOAPIC", sizeEntryIOAPIC},
+		entryISOR:          {"ISOR", sizeEntryISOR},
+		entryNMI:           {"NMI", sizeEntryNMI},
+		entryLAPICOverRide: {"LAPICOverRide", sizeEntryLAPICOverRide},
+		entryLX2APIC:       {"LX2APIC", sizeEntryLX2APIC},
+		entryLX2NMI:        {"X2NMI", sizeEntryLX2NMI},
+	}
+)
+
+// MADT defines the madt struct.
+type MADT struct {
+	acpi.Table
+	// base is the base address of the MADT struct in physical memory.
+	base int64
+}
+
+func init() {
+	corebooters["APIC"] = NewMADT
+}
+
+// rv reverses a slice
+func rv(b []byte) []byte {
+	for i := 0; i < len(b)/2; i++ {
+		j := len(b) - i - 1
+		b[i], b[j] = b[j], b[i]
+	}
+	return b
+}
+
+// NewMADT returns a MADT, from a Table, iff the signature is correct.
+func NewMADT(m acpi.Table) (Corebooter, error) {
```

Comment:
nit: return the specific type, not the interface, so *MADT, error

put a

```
var _ Corebooter = &MADT{}
```

somewhere to make sure that it satisfies the interface at compile time

Reply:
have to return interface or I can't add it to the map above.

Reply:
added var _ etc. etc.

Reply:
I actually got a little curious about this, as my memory was it was pretty common to return the interface name from a New.
grep -r func.*New.*io.Read ~/go
this is fairly common for things that implement, e.g., io.Reader. 
I don't think this declaration is entirely incorrect ...